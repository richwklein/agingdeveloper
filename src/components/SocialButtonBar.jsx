import React from "react";
import {Facebook, GitHub, Instagram, LinkedIn, Pinterest, Public, Reddit, Twitter} from "@mui/icons-material";
import {ButtonGroup, IconButton} from "@mui/material";
import PropTypes from "prop-types";
import {SocialProps} from "../props";
import ExternalLink from "./ExternalLink";

/**
 * A react component for rendering a social media icon.
 * If the name is not provided, a default image is used.
 *
 * @param {SocialIconProps} props
 * @return {React.ReactElement} - The react component.
 */
export const SocialIcon = React.memo(({name=null}) => {
  switch (name.toLowerCase()) {
    case "twitter":
      return <Twitter />;
    case "facebook":
      return <Facebook />;
    case "linkedin":
      return <LinkedIn />;
    case "github":
      return <GitHub />;
    case "instagram":
      return <Instagram />;
    case "pinterest":
      return <Pinterest />;
    case "reddit":
      return <Reddit />;
    default:
      return <Public />;
  }
});

SocialIcon.displayName = "SocialIcon";

/**
 * @typedef SocialIconProps - The props for the social icon.
 * @property {string} [name] - The name of the social site.
 */
SocialIcon.propTypes = {
  name: PropTypes.string,
};

/**
 * A react component for rendering a social media buttons.
 *
 * @param {SocialButtonProps} props
 * @return {React.ReactElement} - The react component.
 */
export const SocialButton = ({social: {name, url}}) => {
  return (
    <IconButton
      component={ExternalLink }
      to={url}
      title={name}
      sx={{color: "primary.main"}}>
      <SocialIcon name={name} />
    </IconButton>
  );
};

/**
 * @typedef SocialButtonProps - The social links button props.
 * @property {SocialProp} social - The social metadata.
 */
SocialButton.propTypes = {
  social: SocialProps,
};

/**
 * A react component for rendering a bar of social media buttons.
 *
 * @param {SocialButtonBarProps} props
 * @return {React.ReactElement} - The react component.
 */
export const SocialButtonBar = ({socials}) => {
  return (
    <ButtonGroup>
      {socials.map((social) => {
        return (
          <SocialButton key={social.name} social={social} />
        );
      })};
    </ButtonGroup>);
};

/**
 * @typedef SocialButtonBarProps - The social links button bar props.
 * @property {SocialProps[]} socials - A list of social metadata.
 */
SocialButtonBar.propTypes = {
  socials: PropTypes.arrayOf(SocialProps).isRequired,
};
export default SocialButtonBar;
