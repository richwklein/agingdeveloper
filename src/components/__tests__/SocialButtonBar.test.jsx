import React from "react";
import {cleanup, render} from "@testing-library/react";
import SocialButtonBar, {SocialIcon} from "../SocialButtonBar";

describe("SocialButtonBar", () => {
  afterEach(cleanup);

  const renderComponent = () => render(
      <SocialButtonBar
        socials={[
          {name: "Twitter", url: "https://twitter.com"},
          {name: "Pinterest", url: "https://pinterest.com"},
        ]}/>,
  );

  test.todo("Implement tests.");

  test("that it renders correctly.", () => {
    const {container} = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("that the icons and titles are correct.", () => {
    const {container, getByTestId} = renderComponent();
    expect(getByTestId("TwitterIcon")).toBeInTheDocument();
    expect(container.firstChild).toContainHTML("title=\"Twitter\"");
    expect(container.firstChild).toContainHTML("href=\"https://twitter.com\"");

    expect(getByTestId("PinterestIcon")).toBeInTheDocument();
    expect(container.firstChild).toContainHTML("title=\"Pinterest\"");
    expect(container.firstChild).toContainHTML("href=\"https://pinterest.com\"");
  });

  test("the default icon is used", () => {
    const {getByTestId} = render(<SocialIcon name="unknown" />);
    expect(getByTestId("PublicIcon")).toBeInTheDocument();
  });
});
