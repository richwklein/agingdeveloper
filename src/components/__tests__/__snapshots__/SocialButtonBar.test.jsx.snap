// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialButtonBar that it renders correctly. 1`] = `
<div
  class="MuiButtonGroup-root MuiButtonGroup-outlined css-edsu1d-MuiButtonGroup-root"
  role="group"
>
  <a
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1wlonk4-MuiTypography-root-MuiLink-root-MuiButtonBase-root-MuiIconButton-root"
    data-link="external"
    href="https://twitter.com"
    rel="noopener noreferrer"
    tabindex="0"
    target="_blank"
    title="Twitter"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="TwitterIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M22.46 6c-.77.35-1.6.58-2.46.69.88-.53 1.56-1.37 1.88-2.38-.83.5-1.75.85-2.72 1.05C18.37 4.5 17.26 4 16 4c-2.35 0-4.27 1.92-4.27 4.29 0 .34.04.67.11.98C8.28 9.09 5.11 7.38 3 4.79c-.37.63-.58 1.37-.58 2.15 0 1.49.75 2.81 1.91 3.56-.71 0-1.37-.2-1.95-.5v.03c0 2.08 1.48 3.82 3.44 4.21a4.22 4.22 0 0 1-1.93.07 4.28 4.28 0 0 0 4 2.98 8.521 8.521 0 0 1-5.33 1.84c-.34 0-.68-.02-1.02-.06C3.44 20.29 5.7 21 8.12 21 16 21 20.33 14.46 20.33 8.79c0-.19 0-.37-.01-.56.84-.6 1.56-1.36 2.14-2.23z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </a>
  <a
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1wlonk4-MuiTypography-root-MuiLink-root-MuiButtonBase-root-MuiIconButton-root"
    data-link="external"
    href="https://pinterest.com"
    rel="noopener noreferrer"
    tabindex="0"
    target="_blank"
    title="Pinterest"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="PinterestIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M9.04 21.54c.96.29 1.93.46 2.96.46a10 10 0 0 0 10-10A10 10 0 0 0 12 2 10 10 0 0 0 2 12c0 4.25 2.67 7.9 6.44 9.34-.09-.78-.18-2.07 0-2.96l1.15-4.94s-.29-.58-.29-1.5c0-1.38.86-2.41 1.84-2.41.86 0 1.26.63 1.26 1.44 0 .86-.57 2.09-.86 3.27-.17.98.52 1.84 1.52 1.84 1.78 0 3.16-1.9 3.16-4.58 0-2.4-1.72-4.04-4.19-4.04-2.82 0-4.48 2.1-4.48 4.31 0 .86.28 1.73.74 2.3.09.06.09.14.06.29l-.29 1.09c0 .17-.11.23-.28.11-1.28-.56-2.02-2.38-2.02-3.85 0-3.16 2.24-6.03 6.56-6.03 3.44 0 6.12 2.47 6.12 5.75 0 3.44-2.13 6.2-5.18 6.2-.97 0-1.92-.52-2.26-1.13l-.67 2.37c-.23.86-.86 2.01-1.29 2.7v-.03z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </a>
</div>
`;

exports[`SocialButtonBar that the icons and titles are correct. 1`] = `
<div
  class="MuiButtonGroup-root MuiButtonGroup-outlined css-edsu1d-MuiButtonGroup-root"
  role="group"
>
  <a
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1wlonk4-MuiTypography-root-MuiLink-root-MuiButtonBase-root-MuiIconButton-root"
    data-link="external"
    href="https://twitter.com"
    rel="noopener noreferrer"
    tabindex="0"
    target="_blank"
    title="Twitter"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="TwitterIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M22.46 6c-.77.35-1.6.58-2.46.69.88-.53 1.56-1.37 1.88-2.38-.83.5-1.75.85-2.72 1.05C18.37 4.5 17.26 4 16 4c-2.35 0-4.27 1.92-4.27 4.29 0 .34.04.67.11.98C8.28 9.09 5.11 7.38 3 4.79c-.37.63-.58 1.37-.58 2.15 0 1.49.75 2.81 1.91 3.56-.71 0-1.37-.2-1.95-.5v.03c0 2.08 1.48 3.82 3.44 4.21a4.22 4.22 0 0 1-1.93.07 4.28 4.28 0 0 0 4 2.98 8.521 8.521 0 0 1-5.33 1.84c-.34 0-.68-.02-1.02-.06C3.44 20.29 5.7 21 8.12 21 16 21 20.33 14.46 20.33 8.79c0-.19 0-.37-.01-.56.84-.6 1.56-1.36 2.14-2.23z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </a>
  <a
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1wlonk4-MuiTypography-root-MuiLink-root-MuiButtonBase-root-MuiIconButton-root"
    data-link="external"
    href="https://pinterest.com"
    rel="noopener noreferrer"
    tabindex="0"
    target="_blank"
    title="Pinterest"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="PinterestIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M9.04 21.54c.96.29 1.93.46 2.96.46a10 10 0 0 0 10-10A10 10 0 0 0 12 2 10 10 0 0 0 2 12c0 4.25 2.67 7.9 6.44 9.34-.09-.78-.18-2.07 0-2.96l1.15-4.94s-.29-.58-.29-1.5c0-1.38.86-2.41 1.84-2.41.86 0 1.26.63 1.26 1.44 0 .86-.57 2.09-.86 3.27-.17.98.52 1.84 1.52 1.84 1.78 0 3.16-1.9 3.16-4.58 0-2.4-1.72-4.04-4.19-4.04-2.82 0-4.48 2.1-4.48 4.31 0 .86.28 1.73.74 2.3.09.06.09.14.06.29l-.29 1.09c0 .17-.11.23-.28.11-1.28-.56-2.02-2.38-2.02-3.85 0-3.16 2.24-6.03 6.56-6.03 3.44 0 6.12 2.47 6.12 5.75 0 3.44-2.13 6.2-5.18 6.2-.97 0-1.92-.52-2.26-1.13l-.67 2.37c-.23.86-.86 2.01-1.29 2.7v-.03z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </a>
</div>
`;
